Stale Caddy config after partial deploy failure: 'already at spec' skips config push

## Summary

When a `yoink up` reconciliation successfully starts new containers and stops old ones, but **fails to push the Caddy config** at the end, a subsequent `yoink up` that finds all containers already at spec skips the Caddy config push entirely — leaving Caddy permanently routing to the now-dead old backends.

## Steps to reproduce

1. Run `yoink up` with a new image tag. New containers start, become healthy, old containers are stopped.
2. At the very end of reconciliation, the Caddy config push fails (e.g. `push Caddy config: HTTP error talking to proxy admin: error sending request for url (http://127.0.0.1:36163/load)`).
3. The workflow fails. Caddy's routing table still points to the old (now stopped) containers.
4. Run `yoink up` again with the same tag. All containers are already running and at spec — yoink reports `already at spec — no-op` for every service.
5. **Caddy config push is skipped.** Caddy continues routing to dead backends.

## Observed behaviour

From a real incident (2026-05-03):

**First deploy (partial failure):**
```
[web] [backtrack-eu-1] started web-cb98d140-0
[web] [backtrack-eu-1] web-cb98d140-0 healthy (1 attempt(s))
[web] [backtrack-eu-1] started web-cb98d140-1
[web] [backtrack-eu-1] web-cb98d140-1 healthy (1 attempt(s))
[web] [backtrack-eu-1] stopped old web-9b0814c5-1
[web] [backtrack-eu-1] stopped old web-9b0814c5-0
[web] [backtrack-eu-1] done — web-cb98d140-0
yoink: reconcile: push Caddy config: HTTP error talking to proxy admin: error sending request for url (http://127.0.0.1:36163/load)
##[error]Process completed with exit code 1.
```

**Second deploy (full no-op):**
```
[yoink-proxy] [backtrack-eu-1] yoink-proxy-ae548533 already at spec — no-op
[api]          [backtrack-eu-1] api-0c6aaa97-0 already at spec — no-op
[api]          [backtrack-eu-1] api-0c6aaa97-1 already at spec — no-op
[web]          [backtrack-eu-1] web-cb98d140-0 already at spec — no-op
[web]          [backtrack-eu-1] web-cb98d140-1 already at spec — no-op
✓ reconciled 6 service(s)
```

Workflow exits **success**, but the site returns **521** because Caddy is still routing to `web-9b0814c5-0/1` (exited).

## Expected behaviour

`yoink up` should always push the Caddy config at the end of every reconciliation, regardless of whether any containers were changed. The push is idempotent — pushing the same config twice is safe and fast. A full no-op reconcile that skips the push means there is no way to recover from a partial failure via a re-run.

## Suggested fix

Unconditionally push the Caddy config at the end of `reconcile()`, even when `changed_services` is empty (all no-ops). The push cost is negligible — one HTTP request to the local admin API. The correctness guarantee is significant.

## Workaround

Roll back to a previous image tag. This forces container replacement, which triggers a Caddy config push and restores routing.

## Environment

- yoink: v0.19.0
- Host: single Hetzner box (backtrack-eu-1)
- Proxy: bundled yoink-proxy (Caddy 2) with Cloudflare origin-pull mTLS

--- a/yoink/src/deploy.rs
+++ b/yoink/src/deploy.rs
@@ -497,6 +497,34 @@ pub async fn reconcile_with_options(
         }
     }
 
+    // Unconditional Caddy push, after every wave has settled. The
+    // per-service path inside `deploy_to_host` already pushes when
+    // it changes containers; this final push covers the recovery
+    // case where every service is already at spec (full no-op
+    // reconcile) but Caddy's routing table is out of date — e.g.
+    // because a previous run died after replacing containers but
+    // before the in-flight push, leaving the operator with no way
+    // to reach a green state without resorting to `--force` or
+    // bumping a tag. The push is idempotent and cheap, so the
+    // redundant calls in the happy path are fine.
+    //
+    // Only run for proxy-enabled configs, and only against hosts
+    // the proxy service targets. `services_filter` is intentionally
+    // ignored: the operator may have run `yoink up --service web`
+    // on a partially-failed reconcile, and we still want Caddy
+    // resynced to actual on-host state.
+    if crate::proxy::proxy_enabled(config)
+        && let Some(proxy_svc) = config
+            .services
+            .iter()
+            .find(|s| matches!(s.kind, Some(crate::config::ServiceKind::Proxy)))
+    {
+        for host_cfg in proxy_svc.applicable_hosts(&config.hosts) {
+            let host = Host::from(host_cfg);
+            push_caddy_config(ops, &host, config, secrets, None).await?;
+        }
+    }
+
     Ok(reports)
 }
 

--- a/yoink/src/deploy.rs
+++ b/yoink/src/deploy.rs
@@ -497,22 +497,12 @@ pub async fn reconcile_with_options(
         }
     }
 
-    // Unconditional Caddy push, after every wave has settled. The
-    // per-service path inside `deploy_to_host` already pushes when
-    // it changes containers; this final push covers the recovery
-    // case where every service is already at spec (full no-op
-    // reconcile) but Caddy's routing table is out of date — e.g.
-    // because a previous run died after replacing containers but
-    // before the in-flight push, leaving the operator with no way
-    // to reach a green state without resorting to `--force` or
-    // bumping a tag. The push is idempotent and cheap, so the
-    // redundant calls in the happy path are fine.
-    //
-    // Only run for proxy-enabled configs, and only against hosts
-    // the proxy service targets. `services_filter` is intentionally
-    // ignored: the operator may have run `yoink up --service web`
-    // on a partially-failed reconcile, and we still want Caddy
-    // resynced to actual on-host state.
+    // Final unconditional Caddy push: recovers from a prior run that
+    // replaced containers and then died before the in-flight push
+    // (#88), since a follow-up reconcile finds every service already
+    // at spec and otherwise never calls `push_caddy_config`.
+    // `services_filter` is ignored on purpose — a scoped retry like
+    // `yoink up --service web` should still resync Caddy.
     if crate::proxy::proxy_enabled(config)
         && let Some(proxy_svc) = config
             .services

